File: app/src/main/java/com/wayfair/themoviedb/feature/searchresults/SearchResultsInteractor.java
package com.wayfair.themoviedb.feature.searchresults;

import com.pubbix.base.BaseInteractor;
import com.wayfair.themoviedb.feature.searchresults.datamodel.SearchResultDataModel;

import java.util.List;

import javax.inject.Inject;

import timber.log.Timber;

public class SearchResultsInteractor extends BaseInteractor<SearchResultsContract.Repository>
        implements SearchResultsContract.Interactor {
    private SearchResultsContract.Presenter presenter;
    private SearchResultsContract.Repository repository;
    private SearchResultsContract.Router router;
    private String searchTerms = "";

    @Inject
    public SearchResultsInteractor(SearchResultsContract.Repository repository,
                                   SearchResultsContract.Router router) {
        this.repository = repository;
        this.repository.setInteractor(this);
        this.router = router;
    }

    @Override
    public void fetchData(String searchTerms) {
        this.searchTerms = searchTerms;
        repository.performMultiSearch(searchTerms);
    }

    @Override
    public void onFetchDataSuccess(List<SearchResultDataModel> searchResultDataModels) {
        presenter.renderSearchResults(searchResultDataModels, searchTerms);
    }

    @Override
    public void onFetchDataFailure(Throwable throwable) {
        Timber.e(throwable);
    }

    @Override
    public void setPresenter(SearchResultsContract.Presenter presenter) {
        this.presenter = presenter;
    }

    @Override
    public void onSearchMovies(String searchTerm) {
        this.searchTerms = searchTerm;
        repository.performMultiSearch(searchTerm);
    }

    @Override
    public void onBackButtonClicked() {
        router.goBack();
    }
}

File: app/src/main/java/com/wayfair/themoviedb/feature/searchresults/SearchResultsInteractor.kt
package com.wayfair.themoviedb.feature.searchresults

import com.pubbix.base.BaseInteractor
import com.wayfair.themoviedb.feature.searchresults.datamodel.SearchResultDataModel
import timber.log.Timber
import javax.inject.Inject

class SearchResultsInteractor @Inject
constructor(private val repository: SearchResultsContract.Repository) : BaseInteractor<SearchResultsContract.Repository>(), SearchResultsContract.Interactor {
    private var presenter: SearchResultsContract.Presenter? = null

    init {
        this.repository.setInteractor(this)
    }

    override fun fetchData(searchTerms: String) {
        repository.performMultiSearch(searchTerms)
    }

    override fun onFetchDataSuccess(searchResultDataModels: List<SearchResultDataModel>) {
        presenter?.renderSearchResults(searchResultDataModels)
    }

    override fun onFetchDataFailure(throwable: Throwable) {
        Timber.e(throwable)
    }

    override fun setPresenter(presenter: SearchResultsContract.Presenter) {
        this.presenter = presenter
    }
}
